### PR TITLE
Make `@particle_input` apply categorization criteria when creating a `ParticleList`

### LINKED ISSUE #2048: `@particle_input` does not apply categorization criteria when creating a `ParticleList`
### Bug description

The `require`, `exclude`, and `any_of` parameters of `@particle_input` do not get enforced when the decorated function gets passed a `ParticleList`.

### Expected outcome

When we use the `require`, `any_of`, and `exclude` parameters with `@particle_input`, we should make them work with `ParticleLike`.

```Python
@particle_input(require="ion", exclude={"isotope"})
def f(x: ParticleLike):
    return x
```

### Minimal complete verifiable example

```Python
>>> from plasmapy.particles import ParticleLike, particle_input
>>> 
>>> @particle_input(require="ion")
... def f(particle: ParticleLike):
...     return particle
...
>>> f(["e-", "e+"])  # should raise an error because these are not ions
ParticleList(['e-', 'e+'])
```

### Package versions

_No response_

### Additional context

When the decorated parameter is named `ion` (for example), `@particle_input` does indeed check that the list contains only ions. 

```pycon
>>> @particle_input
... def g(ion: ParticleLike):
...     return ion
...
>>> g(["e-", "e+"])
plasmapy.particles.exceptions.InvalidIonError: The argument ion = ParticleList(['e-', 'e+']) to g does not correspond to a valid ion.
```

However, there's a little singular vs. plural weirdness in the error message that we could fix.

### PR DESCRIPTION
**Pull request by namurphy**
#2048 describes a problem where particle categorization criteria were not being applied when a `ParticleList` was being supplied to a function decorated with `@particle_input`.

The essense of the error was that a list containing any objects will be interpreted as true, even if the objects the list contains are `False`.

```pycon
>>> bool([False, False, False])
True
```

The improvements in #2648 led to a much cleaner implementation here than I had first attempted (hence the relatively long commit history).

Closes #2048.

### CODE REVIEW

**Comment by namurphy** [comment]
This PR is partially superseded by #2648.  I might close this and open a follow-up PR to add the tests from this PR.  Ultimately, I think #2648 will make this PR (or its follow-up) easier and cleaner.

File: src/plasmapy/particles/particle_class.py [hunk 1]
```diff
@@ -233,7 +233,7 @@ def is_category(
         require: str | Iterable[str] | None = None,
         any_of: str | Iterable[str] | None = None,
         exclude: str | Iterable[str] | None = None,
-    ) -> bool:
+    ) -> bool | list[bool]:
```
Comment (namurphy) [inline src/plasmapy/particles/particle_class.py:236]:
```suggestion
    ) -> bool:
```

File: src/plasmapy/particles/particle_class.py [hunk 2]
```diff
@@ -51,7 +51,7 @@
 _classification_categories: set[str] = {
     "lepton",
     "antilepton",
-    "fermion",
+    "ion",
```
Comment (namurphy) [inline src/plasmapy/particles/particle_class.py:54]:
```suggestion
    "fermion",
```

File: tests/particles/test_particle_collections.py [hunk 3]
```diff
@@ -383,6 +383,12 @@ def test_particle_multiplication(method, particle) -> None:
             {"any_of": {"lepton", "charged"}},
             False,
         ],
+        [
+            ["He-4 1+", "e-"],
+            [],
+            {"any_of": {"ion", "lepton"}},
+            True,
+        ],
```
Comment (namurphy) [inline tests/particles/test_particle_collections.py:391]:
```suggestion
```

File: src/plasmapy/particles/particle_collections.py [hunk 4]
```diff
@@ -346,10 +346,12 @@ def is_category(
...
-        Please refer to the documentation of
-        `~plasmapy.particles.particle_class.Particle.is_category`
-        for information on the parameters and categories, as well as
-        more extensive examples.
+        .. note::
+
+            For parameters to this method and the full list of
+            categories, refer to the documentation for the
+            `~plasmapy.particles.particle_class.Particle.is_category`
+            method of |Particle|.
```
Comment (namurphy) [inline src/plasmapy/particles/particle_collections.py:354]:
```suggestion
        Please refer to the documentation of
        `~plasmapy.particles.particle_class.Particle.is_category`
        for information on the parameters and categories, as well as
        more extensive examples.
```

File: src/plasmapy/particles/decorators.py [hunk 5]
```diff
@@ -493,9 +488,6 @@ def verify_particle_name_criteria(
             else:
                 meets_name_criteria = particle.is_category(**categorization)
```
Comment (namurphy) [inline src/plasmapy/particles/decorators.py:490]:
```suggestion
            if isinstance(particle, Iterable) and not isinstance(particle, str):
                meets_name_criteria = all(meets_name_criteria)  # type: ignore[arg-type]

```

File: tests/particles/test_decorators.py [hunk 6]
```diff
@@ -748,3 +748,31 @@ def function_with_pos_and_var_positional_arguments(
...
+def test_particle_categorization_of_particle_lists(
+    criteria: dict[str, str | Iterable[str]],
+    kwargs: dict[str, ParticleListLike],
+    exception: Exception,
+) -> None:
+    @particle_input(**criteria)  # type: ignore[arg-type]
+    def get_particle(
+        particle: ParticleLike,
+    ) -> Particle | ParticleList | CustomParticle:
+        return particle  # type: ignore[return-value]
```
Comment (namurphy) [inline tests/particles/test_decorators.py:775]: `@particle_input` dynamically transforms objects passed as arguments to functions, it causes problems for _static_ type checkers.  Hence, the `# type: ignore` comments.

**Review by ejohnson-96 — APPROVED** [review]
👍

**Pull request merged**